src/JSONParseErrorException.php

```diff
@@ -0,0 +1,49 @@
+<?php
+/*
+    Copyright 2016 Rustici Software
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+namespace TinCan;
+
+class JSONParseErrorException extends \Exception
+{
+    private static $format = 'Invalid JSON "%s": %s (%d)';
+
+    private $malformedValue;
+    private $jsonErrorNumber;
+    private $jsonErrorMessage;
+
+    public function __construct($malformedValue, $jsonErrorNumber, $jsonErrorMessage, \Exception $previous = null) {
+        $this->malformedValue   = $malformedValue;
+        $this->jsonErrorNumber  = (int) $jsonErrorNumber;
+        $this->jsonErrorMessage = $jsonErrorMessage;
```

Comment:
How do public getter/setters break encapsulation? They provide an exposed interface that can be tested (which IIRC it is), and the internals are still encapsulated. Granted in this case we don't really need setters. Calling the json_* methods in a disconnected location from the triggering json_* functions is effectively a side effect of the constructor which we want to avoid. We've effectively isolated the call to json_decode inside the Trait as a `fromJSON` call which makes it the only use but you could re-use the Exception. One could argue that the Exception class shouldn't be public at all, and should simply be internal to the Trait, but that just makes it more difficult to test.